Implement the RX Depacketizer

levitte opened the pull request:
Implements the design doc/designs/quic-design/rx-depacketizer.md, from #18610.

It relies on functionality from #18676 (ACK Manager implementation), #18795 (QUIC Encoding and decoding functionality), #18870 (QUIC Record Layer design) and #18612 (SSL object refactoring using SSL_CONNECTION object)

levitte commented:
A lot of other functionality is currently not in place and not even in PRs, so there are a lot of comments with the key words "ADD CODE" and "ASSUMPTIONS" that will be amended as development of those components progresses.

Until all of those question marks are resolved, this PR will remain a draft.

levitte commented:
This has now been adapted with #18870 in mind

levitte commented:
I'll do more adaptation to minimize it further, so we can do some very minimal preliminary tests

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+    /* Phase 0: special case packet types that have no payload */
+    if (qpacket->type == QUIC_V1_LONG_PKT_TYPE_RETRY) {
+        /* ADD CODE to handle a retry */
+        goto success;
```
Probably should confirm that `packet_payload_len == 0` in this case?

levitte replied:
That had me look...

Hummmmmmmmmmm

There seems to be a bug in PR #18949, see this comment: https://github.com/openssl/openssl/pull/18949/files#r945600155

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+    ackm_data->is_ack_eliciting = 0;
+
+    while (PACKET_remaining(&pkt) > 0) {
+        if (!ossl_quic_wire_peek_frame_header(&pkt, &frame_type))
```
This looks like it will infinite loop. `ossl_quic_wire_peek_frame_header` only does a peek so does not advance the `PACKET`....so `PACKET_remaining()` will always be ` > 0`

levitte replied:
Right...  unfortunately there's no comfortable function to skip a whole frame regardless of its type.
I can add one.

levitte replied:
> I can add one.

Actually no, that wasn't quite as easy as I thought...  looking more closely, it seems more sane to have handler modes.  Ugly, but better than duplicating the switch into two functions...

levitte replied:
Rewrite done

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+            break;
+        default:
+            /* This frame made the packet ACK eliciting */
+            ackm_data->is_ack_eliciting = 0;
```
Shouldn't that be `= 1`?

levitte replied:
Yup.  Fixed

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+            ack.ack_ranges = ranges;
+            ack.num_ack_ranges = OSSL_NELEM(ranges);
+            /* QUESTION: where do I get the ack delay exponent? */
+            if (!ossl_quic_wire_decode_frame_ack(pkt, ack_delay_exponent,
```
This function is designed to be able to pass a NULL `ack` value initially and return the total_ranges so that you can then allocate sufficient space for them.

levitte replied:
Ah, that wasn't the case last time I looked...  thanks

levitte replied:
Fixed

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+
+        case OSSL_QUIC_FRAME_TYPE_RESET_STREAM: {
+            OSSL_QUIC_FRAME_RESET_STREAM frame_data;
+            SSL *stream;
```
Internally, I would expect something like `OSSL_QUIC_STREAM *`.

levitte replied:
Fixed

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+            if (!ossl_quic_wire_decode_frame_ack(pkt, ack_delay_exponent,
+                                                 &ack, NULL))
+                return 0;
+            if (!ossl_ackm_on_rx_ack_frame(connection->ackm, &ack,
```
`connection` is not declared.

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+                 * ASSUMPTION: ssl_close_connection() closes all the streams
+                 * that are attached to it, then destroys the connection SSL.
+                 */
+                ssl_close_connection(packet->connection);
```
Why do we close the connection in this case? `packet` is not defined here.

levitte replied:
Can't see any more which frame type this is for.  It's `RESET_STREAM` or `STOP_SENDING`, whch have this description:

[`RESET_STREAM`](https://datatracker.ietf.org/doc/html/rfc9000#section-19.4):

> An endpoint that receives a RESET_STREAM frame for a send-only stream MUST terminate the connection with error STREAM_STATE_ERROR.

[`STOP_SENDING`](https://datatracker.ietf.org/doc/html/rfc9000#section-19.5):

> An endpoint that receives a STOP_SENDING frame for a receive-only stream MUST terminate the connection with error STREAM_STATE_ERROR.

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+                 */
+                ssl_close_connection(packet->connection);
+                packet->connection = NULL;
+                ossl_quic_set_state(???, QUIC_STREAM_STATE_ERROR);
```
Probably we need something like `SSLfatal` to close a connection in error and raise appropriate errors.

levitte replied:
Yes.  I've added a temporary static function for now, to be determined what it should be replaced with

mattcaswell commented on ssl/quic/quic_rx_frame.c:
```diff
@@ -0,0 +1,391 @@
+             * exactly what that entails in terms of deallocation of that
+             * packet.
+             */
+            if (!ssl_queue_data(stream, frame_data.data, frame_data.len))
```
I expect the stream data to not be copied into another buffer, but be kept around until it is no longer needed. This will mean we can't release the packet until all streams have finished with the data. We might need to do some ref counting on some quic packet object.

levitte replied:
Agreed...  but how remains to be determined.

levitte replied:
We might want to have a sort of buffer metadata structure with whatever we want (such as a refcount), so we can pass it around without having to worry too much all over the place.  The RX depacketizer code just passes the data on, it shouldn't have to worry about refcounts, but since `ssl_queue_data()` (or whatever that function will end up being called) would grab a reference to that buffer, it might be interested in incrementing the refcount.

mattcaswell replied:
> The RX depacketizer code just passes the data on, it shouldn't have to worry about refcounts, but since ssl_queue_data() (or whatever that function will end up being called) would grab a reference to that buffer, it might be interested in incrementing the refcount.

The RX depacketizer code needs to be aware of this to some extent - since it is that code that calls `ossl_qrx_release_pkt` (at the moment at least). That function call will release the underlying buffer, so it must not be called until everything that is using it has finished with it.

levitte replied:
I was thinking that the the OSSL_QRX_PKT could hold a pointer to that metadata structure instead of directly to the actual data, and that the metadata structure would have pointers to the actual data.  `ossl_qrx_release_pkt()` would then be able to decrement the refcount and all other magic as it likes.

BTW, the `OSSL_QRX_PKT` has a magic opaque pointer...  could it be that we already have the necessary stuff to do that magic?  @hlandau ?

mattcaswell replied:
Not sure why we wouldn't just ref count the `OSSL_QRX_PKT` structure.

mattcaswell replied:
> And I'm not sure why ossl_qrx_release_pkt() can't do that, then...

Because I anticipate the record layer being something that we will eventually make pluggable/fetchable. I think we should keep the interface to the record layer as simple as possible.

levitte replied:
> > And I'm not sure why ossl_qrx_release_pkt() can't do that, then...
> 
> Because I anticipate the record layer being something that we will eventually make pluggable/fetchable. I think we should keep the interface to the record layer as simple as possible.

Do you meant to say that this code would call the plugged funtion directly (i.e. something along the lines of `qrx->release_pkt()`, or would `ossl_qrx_release_pkt()` still be used as a wrapper?  In the latter case, I fail to see the problem.

hlandau replied:
The QRX decrypts entire packets. Once you call `ossl_qrx_release_pkt`, you may no longer reference the decrypted buffer it provides you. (The QRX would be entitled to then free the buffer, in practice it will recycle it for a new packet.)

If there is data in the packet you need to keep for an extended period of time, you need to either copy it (which we are trying to avoid), or avoid releasing the packet until you no longer need to refer to *any* data within it. There is no time limit on this; if all of the QRX's internal buffers are checked out and you try to receive new packets, it will just allocate more.

My assumption here is that there is bookkeeping the RX depacketiser will want to do which means that it will want to define its own internal bookkeeping structure. There is no need to keep the `OSSL_QRX_PKT` structure itself around so long as you remember the `handle` field so you can eventually pass it to `ossl_qrx_release_pkt`. It would probably be logical for the RX depacketiser to keep a refcount to determine when to call `ossl_qrx_release_pkt` if it has complex lifetime requirements.

levitte replied:
Ok...  I was imagining that the RX depacketizer wouldn't have to do that...  Meh...

levitte replied:
I've added bookkeeping code, as a wrapper around the `OSSL_QRX_PKT` handle.

levitte commented:
This currently requires the following additional PRs:

- #18676
- #18949

levitte commented:
This is now ready enough for a real review

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+ * of that packet.
+ */
+static int ssl_queue_data(QUIC_STREAM *stream,
+                          const unsigned char *data, uint64_t data_len)
```
This looks too simplistic to me. I don't think we should be copying data here - just copying pointers to data. That means the underlying buffer needs to remain available for the whole time that the data sits in the stream queue. I suggest we pass the `OSSL_QRX_PKT` object and this function takes a reference to it. Only when the reference drops to 0 would the packet be released and the underlying buffer removed.

levitte replied:
Sure.  Do note that `ssl_queue_data()` is simply a placeholder for whatever the function is actually going to be called, and what parameters it's going to take.  You want a `OSSL_QRX_PKT` parameter?  You got it!

levitte replied:
There, parameter added, and comment enhanced.

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+        return 0;
+
+    rc = ossl_quic_handle_frames(connection, &qpacket);
+    ossl_qrx_release_pkt(GET_CONN_QRX(connection), qpacket.handle);
```
I don't think it is so simple. See my comment above about `ssl_queue_data` and ref counting the `OSSL_QRX_PKT`. In my suggested solution only when the ref count on the `OSSL_QRX_PKT` drops to 0 would we call `ossl_qrx_release_pkt`.

levitte replied:
I assume that `ossl_qrx_release_pkt()` can handle the refcount decrement.
(it could also be modified to take a pointer to `OSSL_QRX_PKT` rather than the mysterious opaque pointer that it holds...  @hlandau?)

levitte replied:
I've addressed the parameter question: https://github.com/openssl/openssl/pull/18949/files#r947918011)

levitte replied:
This has been fixed with a wrapper

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+            ack.ack_ranges = ack_ranges;
+            ack.num_ack_ranges = total_ranges;
+
+            /* QUESTION: where do I get the ack delay exponent? */
```
This is a question for @hlandau I guess.

hlandau replied:
The ACK Delay Exponent is negotiated via a transport parameter (note that each direction can have a different value). The best solution here would probably be to provide a function to let whatever instantiated the RX depacketiser set the value. For testing, 3 is a commonly used value.

levitte replied:
I assume that there would also be a function with which I can *get* that value...  that's the interesting part here.

mattcaswell replied:
If it's negotiated as a transport parameter then presumably you can get it from the QUIC_CONNECTION object somehow (`QUIC_CONNECTION_get_ack_delay_exponent()`). Although what do we use prior to the point that the parameter negotiation is complete? I guess there is a default value that such a function would return if it hasn't been set yet.

levitte replied:
That sounds like an idea.  I can create a placeholder function along those lines, to be replaced when we've made a real one

hlandau replied:
The RFC states that there is a default ACK delay exponent of 3 (e.g. if the transport parameter is not used).

levitte replied:
I've implemented ack delay exponent retrieval

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+# define GET_CONN_STATM(c) NULL
+#endif
+
+/* =====BEGIN TO BE REMOVED===== */
```
BTW, I suggest we adopt a standard mechanism for noting TODO items (like we have for 3.0, and other similar things in the past), e.g.

```` C
/* TODO(QUIC): whatever */
````

levitte replied:
The reason I made this marker is to wrap a whole section that should end up removed, and to make it clear where it begins and ends.  But perhaps something like this?

``` C
/* TODO(QUIC): [BEGIN: TO BE REMOVED] placeholder macros and functions */
...
/* TODO(QUIC): [END: TO BE REMOVED] */

levitte replied:
Done

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+                    OSSL_STATM *statm = GET_CONN_STATM(connection);
+
+                    ssl_close_connection(connection);
+                    ossl_statm_set_error_state(statm, QUIC_STREAM_STATE_ERROR);
```
Why not just make calling `ssl_close_connection` a responsibility of `ossl_statm_set_error_state`? I also think that function should probably raise an error too (much like `SSLfatal` does for TLS).

levitte replied:
How would `ssl_close_connection()` know whether it's being legitimately closed, or closed because of an error?

mattcaswell replied:
My suggestion was to do it the other way around, i.e. make calling `ssl_close_connection` a responsibility of `ossl_statm_set_error_state`. You can just call `ssl_close_connection` directly for cases where it is being closed not because of an error.

levitte replied:
Hmmm, and what should I pass then?  Is every instance of `STREAM_STATE_ERROR` a legitimate reason to close the connection?

levitte replied:
`QUICfatal` it is.

levitte replied:
Done

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+                     * face, and leave it to the state machine what to do with
+                     * the general STREAM_STATE_ERROR state.
+                     */
+                    return 1;
```
Hmmm. This just seems like a fatal error to me. I'm not sure why we would return 1 here. We've given up processing the packet halfway through. Just return 0 and immediately signal a failure of the whole connection, propagating this error up to the return value of `ossl_quic_depacketize`.

levitte replied:
Ok...  trouble is that in the `collect_ack_data` has no idea that this will fail, and if you look here, you'll see that ACK is emitted before the frames are actually processed (`process_frames` run):

https://github.com/openssl/openssl/pull/18838/files#diff-38c7635c72a873d4d53258267676d04ecf7b625177964441ec808e8f3765ce7eR602-R617

Unless I've misunderstood, it must really happen this way, i.e. ACK must be emitted before the frames are actually processed and used.

So as far as I understand, we've a bit of a catch-22, unless you're telling me that the ACK doesn't have to be emited before the frames have been actually used...  that would actually make things easier, as I wouldn't have to process the OSSL_QRX_PKT twice.

levitte replied:
Ok, I imagined it wasn't quite all right to do it like that...  but if that's the case, then fine, I'm game.

levitte replied:
Changed

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+
+                    /*
+                     * There's no point to continue with this packet.  However,
+                     * processing it was a success, so we return with a happy
```
As above. This is a failure IMO.

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+                }
+                break;
+            }
+            }
```
This looks very wrong from a style perspective. But I'm not sure in our coding style how we are supposed to do this.

levitte replied:
This is a consequence of treating `case` lines like labels, i.e. outdent them:

``` C
switch (expr) {
case foo:
    something();
    break;
case bar: {
    int local_num = 0;

    somethingelse(&local_num);
    break;
}
}
```

The other option would be to not outdent the `case` lines:

``` C
switch (expr) {
    case foo:
        something();
        break;
    case bar: {
        int local_num = 0;

        somethingelse(&local_num);
        break;
    }
}
```

In this regard, we're actually a bit undecided (I've seen code where the `case` lines aren't outdented)...  however, the outdented layout was part of the original source reformatting, and is also the current setup in https://github.com/openssl/openssl/blob/master/doc/openssl-c-indent.el

mattcaswell replied:
> we're actually a bit undecided

The coding style is very clear that you are supposed to line up the switch and case statements (as you have done). See chapter 3:

https://www.openssl.org/policies/technical/coding-style.html#chapter-3-placing-braces-and-spaces

But it doesn't really discuss the instance where you have a block inside `{}` as part of that case. This is actually something I don't see much (I don't know if we have other instances of this in the code). I've often just avoided the issue by not using `{}` and just declare any variables in the next block out.

mattcaswell replied:
Perhaps the answer is this:

```` C
switch (expr) {
case foo:
    something();
    break;
case bar: {
        int local_num = 0;

        somethingelse(&local_num);
        break;
    }
}
````
This actually makes sense. Opening a new block with `{` should add an extra level of indent to where you would otherwise be without the new block.

levitte replied:
> (I don't know if we have other instances of this in the code).

To no one's surprise, the answer is a grep away 😉 : 

``` console
$ git grep -E 'case .*: *\{'
crypto/context.c:    case OSSL_LIB_CTX_RAND_CRNGT_INDEX: {
crypto/ec/ecx_meth.c:    case ASN1_PKEY_CTRL_SET1_TLS_ENCPT: {
crypto/evp/evp_enc.c:    case EVP_CTRL_TLS1_1_MULTIBLOCK_AAD: {
crypto/evp/evp_enc.c:    case EVP_CTRL_TLS1_1_MULTIBLOCK_ENCRYPT: {
providers/fips/self_test_kats.c:        case OSSL_PARAM_UNSIGNED_INTEGER: {
providers/fips/self_test_kats.c:        case OSSL_PARAM_UTF8_STRING: {
providers/fips/self_test_kats.c:        case OSSL_PARAM_OCTET_STRING: {
providers/fips/self_test_kats.c:        case OSSL_PARAM_INTEGER: {
```

> I've often just avoided the issue by not using `{}` and just declare any variables in the next block out.

I strongly dislike having to place such fairly block local variables a longer distance away.  It makes code harder to read.

> Perhaps the answer is this:
> 
> ```c
> switch (expr) {
> case foo:
>     something();
>     break;
> case bar: {
>         int local_num = 0;
> 
>         somethingelse(&local_num);
>         break;
>     }
> }
> ```

No, please no.  Ugly!

levitte replied:
I could, of course, add an unnecessary `default:` case, like, for example, here:

https://github.com/openssl/openssl/blob/2f1112b22a826dc8854b41b60a422c987f8ddafb/providers/fips/self_test_kats.c#L181-L188

(without it, you would have seen the same `}` "problem" as you see here)

mattcaswell replied:
Doesn't look ugly to me. Looks quite reasonable IMO and solves the *really* ugly formatting we currently have in this PR.

mattcaswell replied:
> I could, of course, add an unnecessary default: case, like, for example, here:

It doesn't really solve the problem IMO. That "trick" wouldn't work if you needed a block in your default case too.

As @t8m says, this extra indenting (as per my proposal) is exactly what you would do if you put a block in the middle of any other code, e.g.:

```` C
switch (expr) {
case foo:
    something();
    break;
case bar:
    thing();
    {
        int local_num = 0;

        somethingelse(&local_num);
    }
    break;
}
````
So if the above is ok then surely this is too:
```` C
switch (expr) {
case foo:
    something();
    break;
case bar:
    {
        int local_num = 0;

        thing();
        somethingelse(&local_num);
        break;
    }
}
````
And then putting the opening `{` on the same line as the case is just a logical next step.

t8m replied:
Yeah, disagreement is possible, but the current way is IMO outright confusing.

levitte replied:
We might want to formalize this in the coding style guide.  And codify it in [`doc/openssl-c-indent.el`](https://github.com/openssl/openssl/blob/master/doc/openssl-c-indent.el)...

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+            }
+            break;
+        }
+        case OSSL_QUIC_FRAME_TYPE_NEW_TOKEN: {
```
Some `case` statements seem to have a blank line before them, and others don't

levitte replied:
Yes.  I did that with cases that are treated as one and the same, for clearer grouping...  I didn't do that consistently, though.

mattcaswell commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,612 @@
+            case process_frames: {
+                /* ADD CODE to tell the handshake manager that we're done */
+                break;
+            }
+            }
+            break;
+        }
+        }
```
Hmmmmm.....

levitte commented:
cca82837bd8ea3ddfa56717353b97d22cacf1505 implements the alignment suggested here: https://github.com/openssl/openssl/pull/18838#discussion_r946802714

Personally, I *thoroughly* dislike this re-alignment.  It means that statements end up looking weirdly aligned, all because of a brace at the end of a `case` or `default` line, a brace which isn't very visible.  To me, this is *a lot worse*  than the ending double brace that are aligned the same.

t8m commented:
> Personally, I _thoroughly_ dislike this re-alignment. It means that statements end up looking weirdly aligned, all because of a brace at the end of a `case` or `default` line, a brace which isn't very visible. To me, this is _a lot worse_ than the ending double brace that are aligned the same (and mind you, I agree that's not perfect either... but...).

Either we could do it the same way as we require curly brackets for both if and else blocks of if/else statements if one of the blocks needs it. I.E. require extra curly brackets for all cases of a switch where you need one. Or to avoid extra indentation we could allow this:
```C
switch (x) {
case 1:
    stmt;
    break;
case 2: {
    decl somevar;

    stmt-with-somevar;
    another-stmt;
    break;
    }
default: {
    stmt;
    another-stmt;
    }
}
```

levitte commented:
> > Personally, I _thoroughly_ dislike this re-alignment. It means that statements end up looking weirdly aligned, all because of a brace at the end of a `case` or `default` line, a brace which isn't very visible. To me, this is _a lot worse_ than the ending double brace that are aligned the same (and mind you, I agree that's not perfect either... but...).
> 
> Either we could do it the same way as we require curly brackets for both if and else blocks of if/else statements if one of the blocks needs it. I.E. require extra curly brackets for all cases of a switch where you need one. Or to avoid extra indentation we could allow this:
> 
> ```c
> switch (x) {
> case 1:
>     stmt;
>     break;
> case 2: {
>     decl somevar;
> 
>     stmt-with-somevar;
>     another-stmt;
>     break;
>     }
> default: {
>     stmt;
>     another-stmt;
>     }
> }
> ```

Between the choices, I prefer this last one.  The other option would be to stop outdenting case/default lines, but that seems to be a bigger hurdle...

mattcaswell commented:
> ```` c
> switch (x) {
> case 1:
>     stmt;
>     break;
> case 2: {
>     decl somevar;
> 
>     stmt-with-somevar;
>     another-stmt;
>     break;
>     }
> default: {
>     stmt;
>     another-stmt;
>     }
> }
> ````

This is better than what is currently in this PR - but I don't think this is the way to go.

levitte commented:
Re case indentation, I really think that should be a separate and broader coding style discussion.  It seems unlikely that we will get to an agreement here.

Meanwhile, I'm going to modify the cases to the bracing style @t8m proposed, as I find it much less disturbing.

hlandau commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+ * QUIC_CONNECTION it's attached to, and then destroys that QUIC_STREAM
+ * (as well as its SSL object).
+ */
+static int ssl_close_stream(QUIC_STREAM *stream)
```
Yeah, as far as I understand it the closing of each direction is independent. The sending part and the receiving part of a bidirectional stream are finished independently of one another.

levitte replied:
Thanks!  I assume we can re-use the *how* macros used with [shutdown(2)](https://man7.org/linux/man-pages/man2/shutdown.2.html)?

levitte replied:
Done

t8m commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                    return 1;
+                }
+                break;
+                }
```
Although it might look like a bikeshedding issue to outsiders I think we will need a coding-style policy amendment proposal and an OTC vote to resolve this.

t8m replied:
And yeah, this is still ugly although a little less than the original formatting.

levitte replied:
> Although it might look like a bikeshedding issue to outsiders ...

... not just outsiders :wink:
(I'm too passionate not to participate, mind you)

levitte replied:
> And yeah, this is still ugly although a little less than the original formatting.

Thing is that the switch body is hard to get right / consistent, it's a fairly strange construct as the language goes, and many battles have been made over the best way to lay it out.  This will probably not be the last...

Had C been able to have variable declarations interspersed with statements from the very start, we wouldn't have had this discussion...

Me, I'm starting to think that I should put all these "offending" code blocks into separate functions and simply call them.  That would take away the "offending" braces and end this discussion as well.

mattcaswell replied:
That would work for me, and have the benefit of making the function a lot smaller.

levitte replied:
> Me, I'm starting to think that I should put all these "offending" code blocks into separate functions and simply call them. That would take away the "offending" braces and end this discussion as well.

I've decided to go that way.
End of bikeshedding.

t8m commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+    return NULL;
+}
+
+/* ASSUMPTION: ssl_get_stream_tyoe() gets a stream type from a QUIC_STREAM */
```
Typo: ssl_get_stream_type

levitte replied:
Fixed

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+
+    /*
+     * 1. Queuing the data and the parent packet wrapper should happen here,
+     *    a this call with it:
```
Ah, 'a' should be 'and', I think.  Does that make more sense?

levitte replied:
Fixed

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+}
+
+/* ASSUMPTION: ssl_get_stream_tyoe() gets a stream type from a QUIC_STREAM */
+#define SSL_STREAM_TYPE_R       1
+#define SSL_STREAM_TYPE_W       2
+#define SSL_STREAM_TYPE_RW      (SSL_STREAM_TYPE_R|SSL_STREAM_TYPE_W)
```
Sure

levitte replied:
Done

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
+                if ((stream_type & SSL_STREAM_TYPE_R) != 0) {
```
The NULL check you suggested a few lines up is the answer, 'cause I imagine that a stream that has received the stop sending frame can't be accessed any more, i.e. `ssl_get_stream()` should return NULL.

mattcaswell replied:
Not necessarily. The stream probably still exists (perhaps it still has readable data in its buffers - or is still allowed to receive new data event though it is not allowed to send any more). Actually I'm unclear whether it *is* an error to receive a second stop sending frame (e.g. because of a retransmit)...maybe that's allowed and is just ignored??

t8m replied:
It can definitely be received multiple times if the ack for the previous one is lost. It is a connection error if it is received for non-existent or receiving-only stream.

t8m replied:
So this error condition should be probably `(stream_type & SSL_STREAM_TYPE_W) == 0`

levitte replied:
> So this error condition should be probably `(stream_type & SSL_STREAM_TYPE_W) == 0`

Er what???

levitte replied:
Code has moved around, it's better to restart this discussion on the refactored code

t8m commented:
One thing that seems to be missing is to check the allowed frames based on the packet type. I.e., check against this table: https://github.com/openssl/openssl/blob/5f10b1f371469f6ca921337c4d390f05ada6c606/doc/designs/quic-design/tx-packetiser.md#frames

In case of Initial packet type a packet containing such invalid frames may be discarded as a whole. Other violations should cause error being reported and connection close.

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                    = ssl_get_stream(connection, frame_data.stream_id);
+                int stream_type = ssl_get_stream_type(stream);
+
+                ssl_close_stream(stream);
```
Fixed

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
```
Done

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                break;
+            case process_frames: {
+                QUIC_STREAM *stream
+                    = ssl_get_stream(connection, frame_data.stream_id);
```
Yup

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,645 @@
+                 * ASSUMPTION: ssl_get_stream() gets a QUIC_STREAM from a
+                 * QUIC_CONNECTION by stream ID.
+                 */
+                stream = ssl_get_stream(connection, frame_data.stream_id);
```
Yup

levitte commented:
Refactored to avoid the case-and-brace source formatting dispute...  I find the result more appealing, now that it's done.
A bunch of review comments addressed.

levitte commented:
> One thing that seems to be missing is to check the allowed frames based on the packet type. I.e., check against this table: https://github.com/openssl/openssl/blob/5f10b1f371469f6ca921337c4d390f05ada6c606/doc/designs/quic-design/tx-packetiser.md#frames
> 
> In case of Initial packet type a packet containing such invalid frames may be discarded as a whole. Other violations should cause error being reported and connection close.

Good point.  I'll save that for Monday

levitte commented:
> > One thing that seems to be missing is to check the allowed frames based on the packet type. I.e., check against this table: https://github.com/openssl/openssl/blob/5f10b1f371469f6ca921337c4d390f05ada6c606/doc/designs/quic-design/tx-packetiser.md#frames
> > In case of Initial packet type a packet containing such invalid frames may be discarded as a whole. Other violations should cause error being reported and connection close.
> 
> Good point. I'll save that for Monday

Done

paulidale commented:
CIs are relevant.

paulidale commented:
No unit tests.

Should this still be in draft? :smiley:

levitte commented:
> CIs are relevant.

This still depends on #18949, so will obviously not be possible to merge before that one's in.

levitte commented:
I've rebased, and squashed while I was at it.

levitte commented:
I've added pointers to a `OSSL_ACKM` and a `OSSL_QRX` in `QUIC_CONNECTION`, and added `SSL_ctrl()` commands to set them, given a `QUIC_CONNECTION *` passed as an `SSL *`.  This allows test programs to deal with QUIC connections without having to include deeply private headers.

Following that, I've modified `test/quic_record_test.c` to also parse frames and find out if that was successful or not.

This should now be fully reviewable, pending the CI results.

t8m commented:
> I've added pointers to a `OSSL_ACKM` and a `OSSL_QRX` in `QUIC_CONNECTION`, and added `SSL_ctrl()` commands to set them, given a `QUIC_CONNECTION *` passed as an `SSL *`. This allows test programs to deal with QUIC connections without having to include deeply private headers.

I suppose this would be just a temporary thing. IMO we do not want such internal ctrls to expose guts of QUIC implementation. To me it seems less problematic to just add some test utility functions which would include the internal headers to set the structure members directly. But others might have different opinions.

levitte commented:
> > I've added pointers to a `OSSL_ACKM` and a `OSSL_QRX` in `QUIC_CONNECTION`, and added `SSL_ctrl()` commands to set them, given a `QUIC_CONNECTION *` passed as an `SSL *`. This allows test programs to deal with QUIC connections without having to include deeply private headers.
> 
> I suppose this would be just a temporary thing. IMO we do not want such internal ctrls to expose guts of QUIC implementation. To me it seems less problematic to just add some test utility functions which would include the internal headers to set the structure members directly. But others might have different opinions.

The other option I see is to make `QUIC_CONNECTION` "internal" (i.e. defined in a include/internal header) instead of so deeply private.  But that's a lot more invasive, so I'd rather leave that as a future exercise / for another PR.

paulidale reviewed (commented):
I expected to see some code handling the `C` spec here: frames containing only ACKs do not count towards bytes in flight for congestion control purposes.  The depacketiser seems like the natural place to detect this but I don't see any code doing so.

levitte commented on include/internal/quic_rx_depack.h:
```diff
@@ -0,0 +1,16 @@
+# define OSSL_QUIC_RX_DEPACK_H
+
+int ossl_quic_handle_frames(SSL *ssl, OSSL_QRX_PKT *qpacket);
+__owur int ossl_quic_depacketize(SSL *ssl);
```
True, we don't.  Our internal QUIC headers are rather...  topical.  That's understandable, considering that we've split it into individual parts and naturally want to avoid merge conflicts :wink:

levitte commented on include/openssl/ssl.h.in:
```diff
@@ -1484,6 +1484,8 @@ DECLARE_PEM_rw(SSL_SESSION, SSL_SESSION)
 # define SSL_get_max_proto_version(s) \
         SSL_ctrl(s, SSL_CTRL_GET_MAX_PROTO_VERSION, 0, NULL)
 
+/* SSL_ctrl() numbers 32768..INT_MAX are reserved for internal use */
```
Depends.  I've never seen a cmd number that high, so this is de-facto unused until now.

mattcaswell replied:
I don't really understand the added comment text. SSL_METHOD is a private structure, so it is not possible to define a custom ssl_ctrl function. Hence *all* SSL ctrls are internal use.

levitte replied:
???????

`SSL_ctrl()` is a *public* function, and the second argument is usually quite public, such as `SSL_CTRL_GET_MAX_PROTO_VERSION` that's seen two lines up.

mattcaswell replied:
> SSL_ctrl() is a public function, and the second argument is usually quite public, such as SSL_CTRL_GET_MAX_PROTO_VERSION that's seen two lines up.

Yes SSL_ctrl() is a public function, but public consumers have no ability to define new ctrls. So I'm unclear what you mean by "reserved for internal use". Or do you mean there are some ctrls that we define internally that public consumers should not call?

levitte replied:
Ah, now I get what you're saying.  Ok, `SSL_METHOD` not public, so no one outside of our code can come up with any new number anyway.

> Or do you mean there are some ctrls that we define internally that public consumers should not call?

That's what I end up saying, isn't it?

But anyway, I can simply remove that comment.
And no, "reserving" those numbers for hidden ctrls isn't an API break in that case, @paulidale :wink:

levitte replied:
Comment removed

levitte commented on ssl/quic/quic_record_rx_wrap.c:
```diff
@@ -0,0 +1,66 @@
+    return 1;
+}
+
+void ossl_qrx_pkt_wrap_release(OSSL_QRX *qrx, OSSL_QRX_PKT_WRAP *pkt_wrap)
```
Yeah, although this one has a bit of extra semantics

paulidale replied:
I still think it should be called _free.

levitte replied:
Ok, I changed it

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,931 @@
+#include "../ssl_local.h"
+
+/*
+ * ASSUMPTION: the QUIC_CONNECTION structure refers to other related
```
I see no particular reason why these macros should be removed.  The comment might be due for an edit, sure.

The parts `TO BE REMOVED` are things that I assume will end up being implemented elsewhere.  I don't expect that for the local `GET_CONN_` macros

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,931 @@
+
+/* TODO(QUIC): [END: TO BE REMOVED] */
+
+enum depack_modes {
```
You know, when making that sort of claim, a `git grep` helps.

I do find a lot of `typedef`ed enums, that's true.  I don't see the necessity in strictly local enums, though.

As to name endings, I find two ending with `_e` (`bnrand_flag_e` and `ERR_GET_ACTION_e`), four ending with `_en` (`tlsext_index_en`, `downgrade_en`, `ext_return_en` and `test_types_en`), one ending with `_enum` (`pbe_type_enum`), quite a few ending with `_t`, and what seems to be a but-load with no particular ending at all.

So among ways we *mostly* seem to do things, it seems like I've already chosen the common one :wink:

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,931 @@
+{
+    uint64_t max_data = 0;
+
+
```
Ok

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,931 @@
+/*
+ * Helper functions to process different frame types.
+ *
+ * Typically, those that are ACK eliciting will take an OSSL_ACKM_RX_PKT*
```
... or actually say "pointer"

levitte commented:
> I expected to see some code handling the `C` spec here: frames containing only ACKs do not count towards bytes in flight for congestion control purposes. The depacketiser seems like the natural place to detect this but I don't see any code doing so.

Not sure how that's supposed to work in code...  but I guess I should look more closely at #18018

levitte commented:
> > I expected to see some code handling the `C` spec here: frames containing only ACKs do not count towards bytes in flight for congestion control purposes. The depacketiser seems like the natural place to detect this but I don't see any code doing so.
> 
> Not sure how that's supposed to work in code... but I guess I should look more closely at #18018

Nope, that didn't enlighten me...

levitte commented:
It seems that all review comments have been addressed

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,930 @@
+    ackm_data.time = qpacket->time;
+    switch (qpacket->hdr->type) {
+    case QUIC_PKT_TYPE_INITIAL:
+    case QUIC_PKT_TYPE_VERSION_NEG:
```
Considering that this isn't a packet type specified in the RFC, but something synthetic within OpenSSL, I wasn't sure what to do with it.  But ok, you're telling me that it's to be treated as another special case?  Cool, coming up

levitte commented:
All CIs are happy now

t8m commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -0,0 +1,932 @@
+     */
+    if ((stream = ssl_get_stream(connection, frame_data.stream_id)) == NULL)
+        return 0;
+    if (!ssl_queue_data(stream, parent_pkt, frame_data.data, frame_data.len))
+        return 0;
+
+    /* ADD CODE to deal with |frame_data.is_fin| */
```
I suppose the easiest thing would be to leave this up to the stream to handle that flag. I.e., just pass it to the ssl_queue_data() as an argument.

levitte replied:
It's possible.  I'm not ready to assume that's the way things will go, though.

levitte replied:
> > It's possible. I'm not ready to assume that's the way things will go, though.
> 
> Why not? I assume the stream receive buffer would be the only module to know how to handle it, including the possible protocol violations.

I imagine that it could as well be the depacketizer's responsibility to call something like ssl_stream_shutdown() with appropriate arguments if the FIN bit is set.  It depends on how the ssl API ends up being written.

t8m replied:
Well, that call would go to the rcv stream buffers anyway. But it would have to infer the last buffered stream frame somehow. Which might not be the actual last frame as per the offset+length received if the peer violates the protocol. The fin bit must be always associated with the stream frame data to properly detect the violation.

t8m replied:
Actually I see another problem with this prototype - the offset is of course needed by the stream too.

levitte replied:
Ok, I'm passing both.

hlandau commented:
This is necessarily incomplete because lots of things it needs to plug into are not yet completed, but as a framework I think it suffices. It can be changed by subsequent PRs at the same time that the functionality it needs to plug into is added.

I've only two remaining concerns:
- Does this actually need to depend on SSL objects? It could be a separate object, or depend only on `QUIC_CONNECTION`?
- The frame decoding uses two passes. I don't think this is necessary? The ACKM can be informed of a received packet at the end of a single pass which accumulates the is_ack_eliciting flag.

levitte commented:
> I've only two remaining concerns:
> 
>     * Does this actually need to depend on SSL objects? It could be a separate object, or depend only on `QUIC_CONNECTION`?

The way things are currently declared, `QUIC_CONNECTION` is declared as strictly private inside libcrypto, while `SSL` is a well known object.  Either way, it's a fairly natural way to do this, since `SSL` will tend to be the carrying object, as far as I've been able to tell... and the way `QUIC_CONNECTION` is made, it *isa* `SSL` anyway...

>     * The frame decoding uses two passes. I don't think this is necessary? The ACKM can be informed of a received packet at the end of a single pass which both handles frames and accumulates the is_ack_eliciting flag.

This may be a misunderstanding, then.  However, that's how I read https://github.com/openssl/openssl/blob/master/doc/designs/quic-design/quic-ackm.md#on-rx-ack-frame, that the ACKM bits must be fully handled before frame data is processed further.  If that was incorrect, then oh boy I'll happily reduce the code to be one pass only.

t8m commented:
> The way things are currently declared, `QUIC_CONNECTION` is declared as strictly private inside libcrypto, while `SSL` is a well known object.

IMO this was not the intention. The QUIC_CONNECTION should be well public within the libssl of course. We might want to move the declaration somewhere else if the current place is wrong for that purpose.

t8m commented:
The SSL * should be used only for outside APIs as there is (albeit minimal) overhead associated with converting it to QUIC_CONNECTION *.

levitte commented:
> > The way things are currently declared, `QUIC_CONNECTION` is declared as strictly private inside libcrypto, while `SSL` is a well known object.
> 
> IMO this was not the intention. The QUIC_CONNECTION should be well public within the libssl of course. We might want to move the declaration somewhere else if the current place is wrong for that purpose.

I miswrote, I did mean libssl

levitte commented:
> ... We might want to move the declaration somewhere else if the current place is wrong for that purpose.

Wish granted!

t8m reviewed (commented):
Just one nit to fix and I'll approve.

t8m commented:
Also a rebase is needed.

levitte commented on doc/designs/quic-design/rx-depacketizer.md:
```diff
@@ -13,7 +13,7 @@ Main structures
 
 Represented by an `QUIC_CONNECTION` object, defined in
-[`ssl/quic/quic_local.h`](../../../ssl/quic/quic_local.h).
+[`include/internal/quic_ssl.h`](../../../include/internal/quic_local.h).
```
Fixed

levitte commented:
I'll rebase and squash

levitte commented:
> I'll rebase and squash

Done

levitte commented on test/quic_record_test.c:
```diff
@@ -1735,10 +1735,10 @@ static void rx_state_teardown(struct rx_state *s)
 
     if (s->quic_conn != NULL) {
-        ossl_quic_free((SSL *)s->quic_conn);
+        SSL_free((SSL *)s->quic_conn);
```
Sure.  Done

levitte replied:
I backed off that, 'cause the contents of `QUIC_CONNECTION` isn't known here.

hlandau commented:
>This may be a misunderstanding, then. However, that's how I read https://github.com/openssl/openssl/blob/master/doc/designs/quic-design/quic-ackm.md#on-rx-ack-frame, that the ACKM bits must be fully handled before frame data is processed further. If that was incorrect, then oh boy I'll happily reduce the code to be one pass only.

Yeah, it definitely was not my intention to require multiple passes. Go ahead and make this single-pass.

levitte commented:
> ... Go ahead and make this single-pass.

Done.

levitte commented on ssl/quic/quic_rx_depack.c:
```diff
@@ -791,8 +788,18 @@ int ossl_quic_handle_frames(QUIC_CONNECTION *connection, OSSL_QRX_PKT *qpacket)
+        /* ADD CODE to handle a retry */
+        goto success;
+    } else if (qpacket->hdr->type == QUIC_PKT_TYPE_VERSION_NEG) {
+        /* ADD CODE to handle version negotiation */
```
Done with all "ADD CODE" and "ASSUMPTION"

t8m commented:
@hlandau assuming still OK?

hlandau commented:
Still OK

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

levitte commented:
Merged

Thank you all for the reviews!

d5ab48a192d45ec51355ef2a186125961331eb9b Move the QUIC_CONNECTION typedef to internal headers
69ed6760f938975d9cdcc12ec756d58c83ac6b90 Implement the RX Depacketizer
9be2693438756b5f1b789f1b8db76c3b987063dc Extend the RX Depacketizer frame table with what packet types they are valid in
8a163641c1d94c877a46ade8ff2ecefdf5cbbeeb Implement packet type checks in the RX Depacketizer
ecc920b3277311e859282b6d400ba8566d7ea8c1 Modify test/quic_record_test.c to also depacketize
